File: archive/p/pascal/transpose_matrix.pas
program TransposeMatrix;

{$mode objfpc}{$H+}

uses
   Classes,
   Generics.Collections,
   StrUtils,
   SysUtils;

type
   TIntegerList = specialize TList<integer>;

procedure ShowUsage;
begin
   Writeln('Usage: please enter the dimension of the matrix and the serialized matrix');
   Halt(1);
end;

function ParseIntegerList(const S: string): TIntegerList;
var
   Tokens: TStringArray;
   Token: string;
   Value: integer;
begin
   if S.Trim = '' then
      ShowUsage;

   Tokens := S.Split([',']);
   if Length(Tokens) < 2 then
      ShowUsage;

   Result := TIntegerList.Create;
   for Token in Tokens do
   begin
      if not TryStrToInt(Trim(Token), Value) then
      begin
         Result.Free;
         ShowUsage;
      end;
      Result.Add(Value);
   end;
end;

function FormatIntegerList(const List: TIntegerList): string;
var
   i: integer;
begin
   Result := '';
   for i := 0 to List.Count - 1 do
   begin
      if i > 0 then
         Result += ', ';
      Result += IntToStr(List[i]);
   end;
end;

function TransposeMatrix(const Cols, Rows: integer;
const Input: TIntegerList): TIntegerList;
var
   i, j, index: integer;
begin
   Result := TIntegerList.Create;
   Result.Count := Rows * Cols;

   for i := 0 to Pred(Rows) do
      for j := 0 to Pred(Cols) do
      begin
         index := j * Rows + i;
         Result[index] := Input[i * Cols + j];
      end;
end;

var
   Cols, Rows: integer;
   Numbers, Transposed: TIntegerList;
begin
   if ParamCount <> 3 then
      ShowUsage;

   if (not TryStrToInt(ParamStr(1), Cols)) or (not TryStrToInt(ParamStr(2), Rows)) then
      ShowUsage;

   Numbers := ParseIntegerList(ParamStr(3));
   if Numbers.Count <> (Cols * Rows) then
   begin
      Numbers.Free;
      ShowUsage;
   end;

   Transposed := TransposeMatrix(Cols, Rows, Numbers);
   try
      Writeln(FormatIntegerList(Transposed));
   finally
      Numbers.Free;
      Transposed.Free;
   end;
end.

File: archive/v/visual-basic/transpose-matrix.vb
Public Module TransposeMatrix

    Private Sub ShowUsage()
        Console.Error.WriteLine("Usage: please enter the dimension of the matrix and the serialized matrix")
        Environment.Exit(1)
    End Sub


    Private Function ParseIntegerList(input As String) As List(Of Integer)

        If String.IsNullOrWhiteSpace(input) Then
            ShowUsage()
        End If

        Dim numbers As New List(Of Integer)

        For Each token In input.Split(","c, StringSplitOptions.RemoveEmptyEntries Or StringSplitOptions.TrimEntries)

            Dim value As Integer

            If Not Integer.TryParse(token, value) Then
                ShowUsage()
            End If

            numbers.Add(value)

        Next

        Return numbers

    End Function


    Private Function TransposeMatrix(cols As Integer, rows As Integer, input As List(Of Integer)) As List(Of Integer)

        Dim result As New List(Of Integer)(New Integer(rows * cols - 1) {})

        For i As Integer = 0 To rows - 1

            For j As Integer = 0 To cols - 1

                Dim sourceIndex As Integer = i * cols + j
                Dim targetIndex As Integer = j * rows + i

                result(targetIndex) = input(sourceIndex)

            Next

        Next

        Return result

    End Function


    Public Sub Main(args As String())

        If args.Length <> 3 Then
            ShowUsage()
        End If

        Dim cols As Integer
        Dim rows As Integer

        If Not Integer.TryParse(args(0), cols) OrElse Not Integer.TryParse(args(1), rows) Then
            ShowUsage()
        End If

        Dim numbers = ParseIntegerList(args(2))

        If numbers.Count <> cols * rows Then
            ShowUsage()
        End If

        Dim transposed = TransposeMatrix(cols, rows, numbers)

        Console.WriteLine(String.Join(", ", transposed))

    End Sub

End Module